sst/SimengMemInterface.cc

```diff
@@ -0,0 +1,286 @@
+      This often happens on wrongly speculated branches leading to
+      large values. In this case we queue an empty register value
+      which signals an exception. However, wrongly speculated branches
+      lead to a pipeline flush after which execution continues.
+  */
+  if (addrEnd > maxAddrMemory_ || unsignedOverflow_(addrStart, size)) {
+    completedReadRequests_.push_back({target, RegisterValue(), requestId});
+    return;
+  }
+
+  AggregateReadRequest* aggrReq = new AggregateReadRequest(target, requestId);
+  std::vector<StandardMem::Request*> requests =
+      makeSSTRequests<AggregateReadRequest>(aggrReq, addrStart, addrEnd, size);
+  for (StandardMem::Request* req : requests) {
+    mem_->send(req);
+  }
+}
+
+void SimengMemInterface::requestWrite(const MemoryAccessTarget& target,
+                                      const RegisterValue& data) {
+  uint64_t addrStart = target.address;
+  uint64_t size = unsigned(target.size);
+  uint64_t addrEnd = addrStart + size - 1;
+
+  AggregateWriteRequest* aggrReq = new AggregateWriteRequest(target, data);
+  std::vector<StandardMem::Request*> requests =
+      makeSSTRequests<AggregateWriteRequest>(aggrReq, addrStart, addrEnd, size);
+  for (StandardMem::Request* req : requests) {
+    mem_->send(req);
+  }
+}
+
+void SimengMemInterface::tick() { tickCounter_++; }
+
+void SimengMemInterface::clearCompletedReads() {
+  completedReadRequests_.clear();
+}
+
+bool SimengMemInterface::hasPendingRequests() const {
+  return aggregationMap_.size() > 0;
+};
+
+const span<MemoryReadResult> SimengMemInterface::getCompletedReads() const {
+  return {const_cast<MemoryReadResult*>(completedReadRequests_.data()),
+          completedReadRequests_.size()};
+};
+
+void SimengMemInterface::aggregatedReadResponses(
+    AggregateReadRequest* aggrReq) {
+  if (aggrReq->aggregateCount_ != 0) return;
+  std::vector<uint8_t> mergedData;
+  // Loop through the ordered map and merge the data in order inside the
+  // mergedData vector. Also remove entries from the aggregation_map as we loop
+  // through each SST Request id.
+  for (auto itr = aggrReq->responseMap_.begin();
+       itr != aggrReq->responseMap_.end(); itr++) {
+    mergedData.insert(mergedData.end(), itr->second.begin(), itr->second.end());
+    aggregationMap_.erase(itr->first);
+  }
+  // Send the completed read request back to SimEng via the
+  // completed_read_requests queue.
+  const char* char_data = reinterpret_cast<const char*>(&mergedData[0]);
+  completedReadRequests_.push_back(
+      {aggrReq->target,
+       RegisterValue(char_data, uint16_t(unsigned(aggrReq->target.size))),
+       aggrReq->id_});
+
+  // Cleanup
+  aggrReq->responseMap_.clear();
+  delete aggrReq;
+}
+
+void SimengMemInterface::SimengMemHandlers::handle(
+    StandardMem::WriteResp* rsp) {
+  delete rsp;
+}
+
+void SimengMemInterface::SimengMemHandlers::handle(StandardMem::ReadResp* rsp) {
```

Comment:
Is overloading necessary here? Would it be more readable to have two separate function names

Reply:
Also not a fan of "handle" as it isn't very descriptive

Reply:
The overloaded definitions are necessary as this class inherits from `StandardMem::RequestHandler` which exposes two virtual overloaded instances of `handle(...)`. Instance of `SimengMemHandlers` is registered as a polymorphic base in `SimengCoreWrapper`. For this we don't have any control over the naming convention.

Reply:
Sure, thought this may be the case